+
diff --git a/src/inc/gm_backend_cpp.h b/src/inc/gm_backend_cpp.h
index 59eae0c..ea28fd3 100644
--- a/src/inc/gm_backend_cpp.h
+++ b/src/inc/gm_backend_cpp.h
@@ -56,15 +56,63 @@ public:
     virtual void generate_expr_builtin_field(ast_expr_builtin_field* builtinExpr, gm_code_writer& body);
     const char* get_function_name_graph(int methodId);
     const char* get_function_name_nset(int methodId, bool in_parallel = false);
     const char* get_function_name_nseq(int methodId);
     const char* get_function_name_norder(int methodId);
+    const char* get_function_name_map(int methodId, bool in_parallel = false);
+    const char* get_function_name_map_seq(int methodId);
+    const char* get_function_name_map_par(int methodId);
     void add_arguments_and_thread(gm_code_writer& body, ast_expr_builtin* builtinExpr, bool addThreadId);
+    const char* getMapDefaultValueForType(int type);
+    const char* getMapTypeString(int mapType);
+    const char* getAdditionalMapParameters(int mapType);
+
